cleanrl/cql_continuous/sac_cql.py

```diff
@@ -0,0 +1,396 @@
+        else:
+            raise NotImplementedError
+
+        min_qf1_loss, min_qf2_loss = torch.logsumexp(all_q1, 1), torch.logsumexp(all_q2, 1)
+
+        min_qf1_loss = (min_qf1_loss - qf1_a_values).mean() * args.min_q_weight
+        min_qf2_loss = (min_qf2_loss - qf2_a_values).mean() * args.min_q_weight
+
+        if args.with_lagrange:
+            alpha_prime = torch.clamp(log_alpha_prime.exp(), min=0.0, max=10000.0)
+            min_qf1_loss = alpha_prime * (min_qf1_loss - args.lagrange_thresh)
+            min_qf2_loss = alpha_prime * (min_qf2_loss - args.lagrange_thresh)
+
+            alpha_prime_optimizer.zero_grad()
+            alpha_prime_loss = -.5 * (min_qf1_loss + min_qf2_loss)
+            alpha_prime_loss.backward(retain_graph=True)
+            alpha_prime_optimizer.step()
+        
+        cql_qf1_loss = qf1_loss + min_qf1_loss
+        cql_qf2_loss = qf2_loss + min_qf2_loss
+
+        cql_qf_loss = (cql_qf1_loss + cql_qf2_loss) / 2.
+
+        values_optimizer.zero_grad()
+        cql_qf_loss.backward()
+        values_optimizer.step()
+
+        ## Policy network update
+        if global_step % args.policy_frequency == 0: # TD 3 Delayed update support
+            for _ in range(args.policy_frequency): # compensate for the delay by doing 'actor_update_interval' instead of 1
+                pi, log_pi, _ = pg.get_action(s_obs)
+                qf1_pi = qf1.forward(s_obs, pi)
+                qf2_pi = qf2.forward(s_obs, pi)
+                min_qf_pi = torch.min(qf1_pi, qf2_pi).view(-1)
+                policy_loss = ((alpha * log_pi) - min_qf_pi).mean()
+
+                policy_optimizer.zero_grad()
+                policy_loss.backward()
+                policy_optimizer.step()
+            
+            if args.autotune:
+                with torch.no_grad():
+                    _, log_pi, _ = pg.get_action(s_obs)
+                alpha_loss = ( -log_alpha * (log_pi + target_entropy)).mean()
+
+                a_optimizer.zero_grad()
+                alpha_loss.backward()
+                a_optimizer.step()
+                alpha = log_alpha.exp().item()
+            
+        # update the target network
+        if global_step % args.target_network_frequency == 0:
+            for param, target_param in zip(qf1.parameters(), qf1_target.parameters()):
+                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
+            for param, target_param in zip(qf2.parameters(), qf2_target.parameters()):
+                target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
+        
+        # Stop iteration over the dataset
+        if k >= args.epoch_length-1:
+            break
+    
+    eval_stats = test_agent(env, pg)
+    print("[E%4d|I%8d] -- PLoss: %.3f -- QLoss: %.3f -- CQLQloss: %.3f --  Train Mean Ret: %.3f" %
```

Comment:
Do we really read losses from the stdout? if not, maybe just print out global step and episode reward like all other scripts. To make comparison with other algo, maybe each `global_step` could mean a gradient step.

Reply:
Mainly used for debug purpose in the early implementation phase.

Changed to `print(f"global_step={global_step}, episode_reward={avg_eval_return}")` following the other scripts
(actually, `ppo_procgen.py`).